mathjax3-ts/input/tex/FilterUtil.ts

```diff
@@ -0,0 +1,174 @@
+      if (NodeUtil.getProperty(mo, 'fixStretchy')) {
+        let symbol = NodeUtil.getForm(mo);
+        if (symbol && symbol[3] && symbol[3]['stretchy']) {
+          NodeUtil.setAttribute(mo, 'stretchy', false);
+        }
+        const parent = mo.parent;
+        if (!NodeUtil.getTexClass(mo) && (!symbol || !symbol[2])) {
+          const texAtom = options.nodeFactory.create('node', 'TeXAtom', [mo]);
+          parent.replaceChild(texAtom, mo);
+          texAtom.setInheritedAttributes({}, arg.math['display'],
+                                         parent.attributes.get('scriptlevel') as number,
+                                         parent.getProperty('texprimestyle') as boolean);
+        }
+        NodeUtil.removeProperties(mo, 'fixStretchy');
+      }
+    }
+  };
+
+
+  /**
+   * Visitor that removes superfluous attributes from nodes. I.e., if a node has
+   * an attribute, which is also an inherited attribute it will be removed. This
+   * is necessary as attributes are set bottom up in the parser.
+   * @param {MmlNode} mml The node to clean.
+   * @param {ParseOptions} options The parse options.
+   */
+  // TODO (DC): Move this maybe into setInheritedAttributes method?
+  export let cleanAttributes = function(arg: {data: ParseOptions}) {
+    let node = arg.data.root as MmlNode;
+    node.walkTree((mml: MmlNode, d: any) => {
+      let attribs = mml.attributes as any;
+      for (const key of attribs.getExplicitNames()) {
+        if (attribs.attributes[key] === mml.attributes.getInherited(key)) {
+          delete attribs.attributes[key];
+        }
+      }
+    }, {});
+  };
+
+
+  /**
+   * Combine adjacent <mo> elements that are relations (since MathML treats the
+   * spacing very differently)
+   * @param {MmlNode} mml The node in which to combine relations.
+   * @param {ParseOptions} options The parse options.
+   */
+  export let combineRelations = function(arg: {data: ParseOptions}) {
+    for (let mo of arg.data.getList('mo')) {
+      if (mo.getProperty('relationsCombined') || !mo.parent ||
+          (mo.parent && !NodeUtil.isType(mo.parent, 'mrow'))) {
+        continue;
+      }
+      let mml = mo.parent;
+      let m1: MmlNode, m2: MmlNode;
+      let children = mml.childNodes as (MmlNode|TextNode)[];
+      for (let i = 0, m = children.length; i < m; i++) {
+        if (!children[i]) {
+          break;
+        }
+        while (i + 1 < m && (m1 = children[i]) && (m2 = children[i + 1]) &&
+               NodeUtil.isType(m1, 'mo') && NodeUtil.isType(m2, 'mo') &&
+               NodeUtil.getTexClass(m1) === TEXCLASS.REL &&
+               NodeUtil.getTexClass(m2) === TEXCLASS.REL) {
+          m2.setProperty('relationsCombined', true);
+          if (NodeUtil.getProperty(m1, 'variantForm') ===
+              NodeUtil.getProperty(m2, 'variantForm') &&
+              NodeUtil.getAttribute(m1, 'mathvariant') ===
+              NodeUtil.getAttribute(m2, 'mathvariant')) {
```

Comment:
@dpvc Here's a question: What if I have a  stretchy character with `stretchy=false`. Can I combine that with an ordinary relation? E.g., `\rightarrow=`. This could be combined previously, but fails when comparing all explicit attributes.

Reply:
Good question.  Really, the main reason for this is to combine things like `<mo>:</mo><mo>=</mo>` into `<mo>:=</mo>` where the combined version is something that might be in the operator dictionary.  I don't think there are any that include arrows.  I'm inclined to not combine when the attributes differ, even in this case.  If you prefer combining them anyway, then I guess you would need a list of attributes to check, which would need to include things like `mathvariant`, `mathsize`, `mathbackrgound`, `mathcolor`, `href`, `style`, `id`, `class`, `linebreak`, and so on.

Reply:
I have now implemented a filter ([here](https://github.com/mathjax/mathjax-v3/pull/116/commits/51cca744180330c3dd97ccad502175544423f3b8#diff-06bfcc41f513ee6d8a9fffc9bd1d435aR125)) that deals with the following exceptions:
   * `lspace` attribute of left node is ignored.
   * `rspace` attribute of right node is ignored.
   * `stretchy=false` is ignored.

Reply:
OK, that sounds good.  I haven't looked at the changes yet, but do you retain the `stretchy="false"` on the combined `<mo>`, or do you drop it?

Reply:
Everything is retained (not in the pushed version yet). I simply assume that a `stretchy=false` for a non-stretchable relation does not matter. E.g.,
```XML
<mo stretchy="false">→=</mo>
```
But I still need to write a couple of tests to be sure that `lspace`, `rspace` work as well.

Reply:
Sounds good.  That is what I would have recommended as well.